src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1292 @@
+
+        if set_new_cache:
+            await self._set_cached_srv_records(SRV['DOMAINCONTROLLER'], site=ad['site'], results=dcs)
+
+        with ActiveDirectory_LDAP(ad_conf=ad, logger=self.logger, hosts=dcs) as AD_LDAP:
+            ret = AD_LDAP.get_netbios_name()
+
+        if ret and smb['workgroup'] != ret:
+            self.logger.debug(f'Updating SMB workgroup to match the short form of the AD domain [{ret}]')
+            await self.middleware.call('datastore.update', 'services.cifs', smb['id'], {'cifs_srv_workgroup': ret})
+
+        return ret
+
+    @private
+    async def get_kerberos_servers(self):
+        """
+        This returns at most 3 kerberos servers located in our AD site. This is to optimize
+        kerberos configuration for locations where kerberos servers may span the globe and
+        have equal DNS weighting. Since a single kerberos server may represent an unacceptable
+        single point of failure, fall back to relying on normal DNS queries in this case.
+        """
+        ad = await self.config()
+        with ActiveDirectory_DNS(conf=ad, logger=self.logger) as AD_DNS:
+            krb_kdc = AD_DNS.get_n_working_servers(SRV['KERBEROSDOMAINCONTROLLER'], 3)
+            krb_admin_server = AD_DNS.get_n_working_servers(SRV['KERBEROS'], 3)
+            krb_kpasswd_server = AD_DNS.get_n_working_servers(SRV['KPASSWD'], 3)
+        kdc = [i['host'] for i in krb_kdc]
+        admin_server = [i['host'] for i in krb_admin_server]
+        kpasswd = [i['host'] for i in krb_kpasswd_server]
+        for servers in [kdc, admin_server, kpasswd]:
+            if len(servers) == 1:
+                return None
+
+        return {'krb_kdc': kdc, 'krb_admin_server': admin_server, 'krb_kpasswd_server': kpasswd}
+
+    @private
+    def set_ntp_servers(self):
+        """
+        Appropriate time sources are a requirement for an AD environment. By default kerberos authentication
+        fails if there is more than a 5 minute time difference between the AD domain and the member server.
+        If the NTP servers are the default that we ship the NAS with. If this is the case, then we will
+        discover the Domain Controller with the PDC emulator FSMO role and set it as the preferred NTP
+        server for the NAS.
+        """
+        ntp_servers = self.middleware.call_sync('system.ntpserver.query')
+        default_ntp_servers = List(filter(lambda x: 'freebsd.pool.ntp.org' in x['address'], ntp_servers)) 
+        if len(ntp_servers) != 3 and len(default_ntp_servers) != 3:
+            return
+
+        ad = self.middleware.call_sync('activedirectory.config')
+        with ActiveDirectory_DNS(conf=ad, logger=self.logger) as AD_DNS:
+            pdc = AD_DNS.get_n_working_servers(SRV['PDC'], 1)
+        self.middleware.call_sync('system.ntpserver.create', {'address': pdc[0]['host'], 'prefer': True})
+
+    @private
+    async def get_site(self):
```

Comment:
And here, ActiveDirectory_DNS being synchronous

Reply:
These have now been changed to methods